server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -685,6 +686,21 @@ protected Node(final Environment initialEnvironment,
         }
     }
 
+    /**
+     * @return A function that can be used to determine the requested REST compatible version
+     */
+    private RestCompatibility getRestCompatibleFunction(){
+        List<RestCompatibility> restCompatibilityPlugins = pluginsService.filterPlugins(RestCompatibility.class);
+        RestCompatibility restCompatibleFunction = RestCompatibility.CURRENT_VERSION;
+        if (restCompatibilityPlugins.size() > 1) {
+            throw new IllegalStateException("Only one rest compatibility plugin is allowed");
+        } else if (restCompatibilityPlugins.size() == 1){
+            restCompatibleFunction =
+                restCompatibilityPlugins.get(0);
```

Comment:
This really isn't a function (ie method reference) it is an instance of the plugin which satisfies the functional interface. I don't think we should pass the plugin instance around.  

I think you can just change this to
```            
restCompatibleFunction = (a,b,c) ->
                restCompatibilityPlugins.get(0).getCompatibleVersion(a,b,c);
``` 
so that you are passing around the method reference. However, it is a bit confusing to return a method reference that satifies the same interface as the object.  Maybe two interfaces (one functional like you have, and another a marker interface to filter the plugins) would read better ?  Maybe a functional interface `CompatibleVersion` (instead of RestCompatibility) , and marker interface `RestCompatibilityPlugin` ?

Reply:
I gave it a go, let me know if this looks any better.
There is a CompatibleVersion field on rest request, but there is a `getCompatibleVersionFunction` to return it in order to implement a copying constructor.
There is also a `getCompatibleVersion` function on the `RestRequest` object itself which uses the `CompatibleVersion` field.

It might be a bit confusing, WDYT?

Reply:
also I am not sure about the marker interface. We still have to define a function on it in order to use it, right?